src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+  auto xloc = Matrix<dcomplex>(4,nmode); // Values of x at the processor interface
+
+  BoutReal kwaveFactor = 2.0 * PI / coords->zlength();
+
+  // Should we store coefficients?
+  store_coefficients = not (inner_boundary_flags & INVERT_AC_GRAD);
+  store_coefficients = store_coefficients && not (outer_boundary_flags & INVERT_AC_GRAD);
+  store_coefficients = store_coefficients && not (inner_boundary_flags & INVERT_SET);
+  store_coefficients = store_coefficients && not (outer_boundary_flags & INVERT_SET);
+
+  // Setting the width of the boundary.
+  // NOTE: The default is a width of 2 guard cells
+  int inbndry = localmesh->xstart, outbndry=localmesh->xstart;
+
+  // If the flags to assign that only one guard cell should be used is set
+  if((global_flags & INVERT_BOTH_BNDRY_ONE) || (localmesh->xstart < 2))  {
+    inbndry = outbndry = 1;
+  }
+  if (inner_boundary_flags & INVERT_BNDRY_ONE)
+    inbndry = 1;
+  if (outer_boundary_flags & INVERT_BNDRY_ONE)
+    outbndry = 1;
+
+  /* Allocation for
+  * bk   = The fourier transformed of b, where b is one of the inputs in
+  *        LaplaceIPT::solve()
+  * bk1d = The 1d array of bk
+  * xk   = The fourier transformed of x, where x the output of
+  *        LaplaceIPT::solve()
+  * xk1d = The 1d array of xk
+  */
+  auto bk = Matrix<dcomplex>(ncx, ncz / 2 + 1);
+  auto bk1d = Array<dcomplex>(ncx);
+  auto xk = Matrix<dcomplex>(ncx, ncz / 2 + 1);
+  auto xk1d = Matrix<dcomplex>(ncz/2+1,ncx);
+
+  // Error interpolated onto a finer grid
+  auto fine_error = Matrix<dcomplex>(4,ncz/2+1);
+
+  ///SCOREP_USER_REGION_END(initvars);
+  ///SCOREP_USER_REGION_DEFINE(initloop);
+  ///SCOREP_USER_REGION_BEGIN(initloop, "init xk loop",SCOREP_USER_REGION_TYPE_COMMON);
+
+  // Initialise xk to 0 as we only visit 0<= kz <= maxmode in solve
+  for (int ix = 0; ix < ncx; ix++) {
+    // TODO This looks wrong: should it be kz<ncz?
+    // This came from serial tri
+    for (int kz = maxmode + 1; kz < ncz / 2 + 1; kz++) {
+      xk(ix, kz) = 0.0;
+    }
+  }
+  ///SCOREP_USER_REGION_END(initloop);
+  ///SCOREP_USER_REGION_DEFINE(fftloop);
+  ///SCOREP_USER_REGION_BEGIN(fftloop, "init fft loop",SCOREP_USER_REGION_TYPE_COMMON);
+
+  /* Coefficents in the tridiagonal solver matrix
+  * Following the notation in "Numerical recipes"
+  * avec is the lower diagonal of the matrix
+  * bvec is the diagonal of the matrix
+  * cvec is the upper diagonal of the matrix
+  * NOTE: Do not confuse avec, bvec and cvec with the A, C, and D coefficients
+  *       above
+  */
+  auto avec = Matrix<dcomplex>(nmode,ncx);
+  auto bvec = Matrix<dcomplex>(nmode,ncx);
+  auto cvec = Matrix<dcomplex>(nmode,ncx);
+  auto bcmplx = Matrix<dcomplex>(nmode,ncx);
+
+  BOUT_OMP(parallel for)
+  for (int ix = 0; ix < ncx; ix++) {
+    /* This for loop will set the bk (initialized by the constructor)
+    * bk is the z fourier modes of b in z
+    * If the INVERT_SET flag is set (meaning that x0 will be used to set the
+    * bounadry values),
```

Comment:
Hmm, it is... it's not my comment though :flushed:   I've structured this solver to be the same as `serial_tri.cxx`, but where serial_tri calls the Thomas algorithm (locally, because it's serial code), I call the Thomas algorithm plus multigrid.  But the code around that is shared and could be separated out, since it's used by this, serial tri and probably other solvers.